pkg/csi/service/fcd/controller.go

```diff
@@ -72,15 +75,133 @@ func (c *controller) CreateVolume(
+		return nil, status.Errorf(codes.Internal, "Volume name is a required parameter.")
+	} else if len(secrets[AttributeFirstClassDiskParentType]) == 0 {
+		return nil, status.Errorf(codes.Internal, "Volume parameter %s is a required parameter.", AttributeFirstClassDiskParentType)
+	} else if len(secrets[AttributeFirstClassDiskParentName]) == 0 {
+		return nil, status.Errorf(codes.Internal, "Volume parameter %s is a required parameter.", AttributeFirstClassDiskParentName)
+	}
+
+	//TODO: zones are currently unimplemented. supports single VC/DC only.
+	// Please see function for more details
+	zone := "TODO"
+	discoveryInfo, err := c.connMgr.WhichVCandDCByZone(ctx, zone)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "Failed to retrieve VC/DC based on zone %s. Err: %v", zone, err)
+	}
+
+	firstClassDisk, err := discoveryInfo.DataCenter.GetFirstClassDisk(
+		ctx, datastoreName, datastoreType, volName, vclib.FindFCDByName)
+	if err == nil {
+		log.Warningf("Volume with name %s already exists. Checking for similar parameters.", volName)
+
+		if firstClassDisk.Config.CapacityInMB != volSizeMB {
+			return nil, status.Errorf(codes.Internal,
+				"Volume already exists but requesting different size. Existing %d != Requested %d",
+				firstClassDisk.Config.CapacityInMB, volSizeMB)
+		}
+	} else {
+		err = discoveryInfo.DataCenter.CreateFirstClassDisk(ctx, datastoreName, datastoreType, volName, volSizeMB)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "CreateFirstClassDisk failed. Err: %v", err)
+		}
+
+		firstClassDisk, err = discoveryInfo.DataCenter.GetFirstClassDisk(
+			ctx, datastoreName, datastoreType, volName, vclib.FindFCDByName)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "GetFirstClassDiskByName(%s) failed. Err: %v", volName, err)
+		}
+	}
+
+	attributes := make(map[string]string)
+	attributes[AttributeFirstClassDiskType] = FirstClassDiskTypeString
+	attributes[AttributeFirstClassDiskName] = firstClassDisk.Config.Name
+	attributes[AttributeFirstClassDiskParentType] = string(firstClassDisk.ParentType)
+	if firstClassDisk.ParentType == vclib.TypeDatastoreCluster {
+		attributes[AttributeFirstClassDiskParentName] = firstClassDisk.StoragePodInfo.Summary.Name
+		attributes[AttributeFirstClassDiskOwningDatastore] = firstClassDisk.DatastoreInfo.Info.Name
+	} else {
+		attributes[AttributeFirstClassDiskParentName] = firstClassDisk.DatastoreInfo.Info.Name
+	}
+
+	resp := &csi.CreateVolumeResponse{
+		Volume: &csi.Volume{
+			VolumeId:      firstClassDisk.Config.Id.Id,
+			CapacityBytes: int64(units.FileSize(firstClassDisk.Config.CapacityInMB * MbInBytes)),
+			VolumeContext: attributes,
+			//TODO: ContentSource?
+		},
+	}
+
+	return resp, nil
 }
 
 func (c *controller) DeleteVolume(
 	ctx context.Context,
 	req *csi.DeleteVolumeRequest) (
 	*csi.DeleteVolumeResponse, error) {
 
-	return nil, nil
+	secrets := req.GetSecrets()
+
+	//check for required parameters
+	if secrets == nil {
```

Comment:
secrets should still not be a required field for us. We're not going to use it, as all the credentials we need will be part of configuring the plugin.

I definitely see a problem, though, so want to mention here again. The CSI spec seems to assume that knowing the volume ID is enough to delete it. For us, I don't think that's the case (at least, not currently). I"m not sure how to handle this if it's the case that we need to know the parent type and parent name -- nothing in the CO is going to pass that to us, and it's not going to be in `secrets` either.